### DISCUSSION TITLE
Temperature difference controller

### DISCUSSION THREAD

**Discussion by der-seemann** [arendst/Tasmota#19702] [Q&A]
I have an aliexpress smart plug with ESP8285N08 running Tasmota and added two DS18B20 temperatur sensors.
I want to control the relay by temperatur difference. Is there a way to do this locally by tasmota without the need of a smart-home-server?

The plug will switch on a fan to vent my basement - but only if its colder outside than inside to keep the basement cool and dry.

**Comment by barbudor** [comment]
You may try using rules with USE_EXPRESSION
Or try to learn Scripting-Language

In the 1st case, you'll need to compile your own version
In the 2nd case there is an "unofficial" tasmota-scripting binary

**Reply by der-seemann** [reply]
well, i got a working rule, but let's move on to scripting, as it has the ability to modify variables by WebUI.

The problem: i cant really set the variables. They show up, but i cant change them. any idea why?

Thats my script right now:


```
>D
; Declare variables
t1=0            ; Indoor temperature
t2=0            ; Outdoor temperature
delta_t=0       ; Temperature difference

p:delta_t_on=4  ; Temperature difference threshold for turning on
p:delta_t_off=2 ; Temperature difference threshold for turning off
p:min_temp=10   ; Minimum temperature for outdoor trigger

; Define temperature sensors
>T
t1=DS18B20_1#Temperature
t2=DS18B20_2#Temperature

; Display temperature information
>S
delta_t=t1-t2
print Indoor Temperature (t1) = %t1%°C
print Outdoor Temperature (t2) = %t2%°C
print Temperature Difference (delta_t) = %delta_t%°K

; Control power based on temperature difference
if delta_t<delta_t_off or t2>min_temp
then
    =>power1 0 ; Turn off power
endif

if delta_t>delta_t_on and t2>min_temp
then
    =>power1 1 ; Turn on power
endif

; Configure web UI to display and edit variables
>W
Indoor Temperature (t1) = %t1%°C
Outdoor Temperature (t2) = %t2%°C
Temperature Difference (delta_t) = %delta_t%°K
Set var = %min_temp% %delta_t_on% %delta_t_off%
nm(-10 20 1 min_temp "Min Temperature / Frost Protection")
nm(0 10 1 delta_t_on "Temperature Difference On")
nm(0 10 1 delta_t_off "Temperature Difference Off")

```

**Reply by der-seemann** [reply]
found the problem. The variables can not contain _ in order to work with nm(

**Reply by der-seemann** [reply]
this is working now :-)

```
; Declare variables
indoorTemp=0           ; Indoor temperature
outdoorTemp=0          ; Outdoor temperature
tempDiff=0             ; Temperature difference

p:tempDiffOn=4.0       ; Temperature difference threshold for turning on
p:tempDiffOff=2.0      ; Temperature difference threshold for turning off
p:minTemp=10.0         ; Minimum temperature for outdoor trigger

; Define temperature sensors
>T
indoorTemp=DS18B20_1#Temperature
outdoorTemp=DS18B20_2#Temperature

; Calculate temperature difference
tempDiff=indoorTemp-outdoorTemp

; Control power based on temperature difference
if tempDiff>tempDiffOn 
and outdoorTemp>minTemp
then
    =>power1 1 ; Turn on power
endif

if tempDiff<tempDiffOff 
or outdoorTemp<minTemp
then
    =>power1 0 ; Turn off power
endif

; Configure web UI to display and edit variables
>W
Indoor Temperature (indoorTemp)=%indoorTemp%°C
Outdoor Temperature (outdoorTemp)=%outdoorTemp%°C
Temperature Difference (tempDiff)=%tempDiff%°K

; Set numeric variables for web UI
nm(-10 20 1 minTemp "Min Temperature / Frost Protection")
nm(0 10 1 tempDiffOn "Temperature Difference On")
nm(0 10 1 tempDiffOff "Temperature Difference Off")

```

**Reply by barbudor** [reply]
Impressive learning curve
In 2 days you've tested Rules and learned Scripting-Language
👏

**Comment by lalo-uy** [comment]
https://tasmota.github.io/docs/Rules/#simple-thermostat-example

**Reply by der-seemann** [reply]
The solar example did something, but not exactly as needed.
So I modified it, but for some reason it stoppt working now at all.

here is my current rule:
`
Rule1
  ON DS18B20-1#temperature DO
    event temp_in=%value%
  ENDON

  ON DS18B20-2#temperature DO
    event temp_out=%value%
  ENDON

  ON event#temp_out-temp_in<=-4 AND event#temp_out>=2 DO
    Power1 1 
  ENDON

  ON event#temp_out-temp_in>-4 OR event#temp_out<2 DO
    Power1 0 
  ENDON 
`

**Reply by sfromis** [reply]
You cannot do arithmetic (the subtraction) or compound logical expressions (AND/OR) in the trigger part to the left of `do`. Triggers is to match a single updating value to another value.

**Reply by barbudor** [reply]
That's why I the 1st answer I suggested Rules with USE_EXPRESSION
Or Scripting-Language

**Reply by der-seemann** [reply]
I have a rule that is working now, just in case anyone stumbles over this and is interested:
its easy to set up, but has limitations like fixed minimum temperature and fixed hysteresis
```
Rule1
ON System#Boot DO mem3 2 ENDON
ON DS18B20-1#temperature DO event t1_in=%value% ENDON
ON DS18B20-2#temperature DO event t2_out=%value% ENDON
ON event#t1_in>%mem3% DO var1 1 ENDON
ON event#t1_in<=%mem3% DO var1 0 ENDON
ON event#t2_out DO Backlog var2 %value% ENDON
ON event#t1_in>%var2% DO Power1 %var1% ENDON
ON event#t1_in<%var2% DO Power1 0 ENDON
```
